pool.go

```diff
@@ -2,165 +2,116 @@ package viamrtsp
-	p.logger.Debug("newFunc for pool was called!")
-	avFrame, err := newAVFrameWrapper()
-	if err != nil {
-		p.logger.Errorf("Failed to allocate AVFrame in frame pool: %v", err)
-		return nil
-	}
-	p.newCount++
-	return avFrame
-}
-
 func (p *framePool) get() *avFrameWrapper {
-	p.mu.RLock()
-	defer p.mu.RUnlock()
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-	if len(p.items) == 0 {
-		return nil
+	if len(p.frames) == 0 {
+		p.logger.Debug("No frames available in pool. Constructing new frame!")
+		frame, err := newAVFrameWrapper()
+		if err != nil {
+			p.logger.Errorf("Failed to allocate AVFrame in frame pool: %v", err)
+			return nil
+		}
+		frame.refs.Store(1)
+		p.newCount++
+		return frame
 	}
 
-	item := p.items[0]
-	p.items = p.items[1:]
+	lastIndex := len(p.frames) - 1
+	frame := p.frames[lastIndex]
+	p.frames = p.frames[:lastIndex]
 
-	p.logger.Debugf("Item was gotten from the pool. Len now: %d", len(p.items))
+	p.logger.Debugf("Item was gotten from the pool. Len now: %d", len(p.frames))
 	p.getCount++
 
-	item.frameWrapper.isInPool.Store(false)
-	return item.frameWrapper
+	frame.isInPool.Store(false)
+	frame.refs.Add(1)
+	return frame
 }
 
 func (p *framePool) put(frame *avFrameWrapper) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-
-	p.items = append(p.items, poolItem{frameWrapper: frame, lastAccess: time.Now()})
-	p.logger.Debugf("Item was put in pool. Len now: %d", len(p.items))
-	p.putCount++
-
-	frame.isInPool.Store(true)
-}
-
-func (p *framePool) safelyPut(frame *avFrameWrapper) {
-	p.logger.Debug("Trying to safely put frame back into pool.")
+	p.logger.Debug("Trying to put frame back into pool.")
 
-	isFreed := frame.isFreed.Load()
-	isBeingServed := frame.isBeingServed.Load()
-	isInPool := frame.isInPool.Load()
-
-	if !isFreed && !isBeingServed && !isInPool {
-		p.put(frame)
-	} else {
-		p.logger.Debugf("Frame was already freed (%t) or is currently being served (%t) or is already in the pool (%t). Cannot put.",
-			isFreed, isBeingServed, isInPool,
-		)
+	if frame.isFreed.Load() {
+		p.logger.Warn("Frame was already freed. Cannot put.")
+		return
+	}
+	if frame.isInPool.Load() {
+		p.logger.Warn("Frame is already in pool. Cannot put")
+		return
+	}
+	if frame.isBeingServed.Load() {
+		p.logger.Debug("Frame is currently being served. Cannot put")
 	}
-}
-
-func (p *framePool) cleanupRoutine(interval time.Duration) {
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
 
-	for {
-		select {
-		case <-ticker.C:
-			p.cleanup()
-		case <-p.stopCleaning:
-			return
-		}
+	refs := frame.refs.Add(-1)
+	if refs < 0 {
+		panic("deref at 0 refs")
+	} else if refs > 0 {
+		p.logger.Debugf("Frame has %d refs, but must have 0 to be put back into pool.", refs)
+		return
 	}
-}
 
-func (p *framePool) cleanup() {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	if len(p.items) <= p.preseededCount {
-		return
+	if len(p.frames) >= p.maxNumFrames {
+		// Cleanup logic: free the last item and truncate the slice
+		lastFrame := p.frames[len(p.frames)-1]
```

Comment:
What I'm reading here is:
* When a frame is deref'ed to a count of 0 and the pool is "full" is ready to use items
  * delete the last frame in the pool
  * and add the deref'ed frame back to the pool

Why do that dance of replacing the last pooled item with the "put" one? Instead of just free'ing the "put" one and calling it a day?

Reply:
Good call; they are for sure equivalent

Reply:
Done